### PR TITLE
Determine whether a stackalloc expression is target-typed when it is bound (not later).

### LINKED ISSUE #40197: Fragile logic in classification of Stackalloc conversion
```
        public override Conversion GetStackAllocConversion(BoundStackAllocArrayCreation sourceExpression, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
        {
            if (sourceExpression.Syntax.IsLocalVariableDeclarationInitializationForPointerStackalloc())
            {
```

We should not be checking syntax here, all required information should be available from bound nodes.

### PR DESCRIPTION
**Pull request by gafter**
This change documents and adopts the design guideline that a bound node shall capture all semantically relevant information from syntax.
Fixes #40197

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs [hunk 1]
```diff
@@ -55,6 +55,11 @@ internal bool NeedsToBeConverted()
                 case BoundKind.TupleLiteral:
                 case BoundKind.UnconvertedSwitchExpression:
                     return true;
+                case BoundKind.StackAllocArrayCreation:
+                    // A BoundStackAllocArrayCreation is given a null type when it is in a
+                    // syntactic context where it could be either a pointer or a span, and
+                    // in that case it requires conversion to one or the other.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpression.cs]: Please capture this information in the BoundNodes.xml above the Type property definition.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -3516,26 +3522,56 @@ private TypeSymbol GetStackAllocType(SyntaxNode node, TypeWithAnnotations elemen
...
+            // Is this a context in which a stackalloc expression could be converted to the corresponding pointer
+            // type? The only context that permits it is the initialization of a local variable declaration (when
+            // the declaration appears as a statement or as the first part of a for loop).
+            static bool isStackallocTargetTyped(SyntaxNode node)
+            {
+                Debug.Assert(node != null);
+
+                SyntaxNode equalsValueClause = node.Parent;
+
+                if (!equalsValueClause.IsKind(SyntaxKind.EqualsValueClause))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: > if (!equalsValueClause.IsKind(SyntaxKind.EqualsValueClause)) [](start = 15, length = 61)

Could it be parenthesized?

Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: No

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: > No 

Do we have a test for a scenario like that?

---
In reply to: [359103173](https://github.com/dotnet/roslyn/pull/40457#discussion_r359103173) [](ancestors = 359103173)

Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Yes.  Customers depend on it.  See for example Nikov's latest comment on Twitter yesterday.

---
In reply to: [359106011](https://github.com/dotnet/roslyn/pull/40457#discussion_r359106011) [](ancestors = 359106011,359103173)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 3]
```diff
@@ -4816,10 +4816,7 @@ unsafe public class Test
             CreateCompilation(test, options: TestOptions.UnsafeDebugDll).VerifyDiagnostics(
                 // (4,14): error CS0518: Predefined type 'System.Span`1' is not defined or imported
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: >// (4,14): error CS0518: Predefined type 'System.Span`1' is not defined or imported [](start = 16, length = 83)

Would this line use System.Span if it was available?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: > Yes.  Customers depend on it.  See for example Nikov's latest comment on Twitter yesterday.

Could you please point to the existing unit-test?

---
In reply to: [359107068](https://github.com/dotnet/roslyn/pull/40457#discussion_r359107068) [](ancestors = 359107068,359106011,359103173)

Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: `StackAllocInitializerTests.StackAllocInParenthesis`

---
In reply to: [359107823](https://github.com/dotnet/roslyn/pull/40457#discussion_r359107823) [](ancestors = 359107823,359107068,359106011,359103173)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: It uses it even though it is not available.

---
In reply to: [359107468](https://github.com/dotnet/roslyn/pull/40457#discussion_r359107468) [](ancestors = 359107468)

**Comment by gafter** [comment]
@AlekseyTs Do you have any other comments?

@dotnet/roslyn-compiler May I have a second review please?

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 4]
```diff
@@ -1764,6 +1764,10 @@
   </Node>
 
   <Node Name="BoundStackAllocArrayCreation" Base="BoundExpression">
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: >BoundStackAllocArrayCreation [](start = 14, length = 28)

This isn't primarily related to the change in this PR, but it looks like this node doesn't follow the guideline that only leaf nodes have kinds.

File: docs/compilers/Design/Bound Node Design.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿Bound Node Design
+=================
+
+This document discusses design principles for the bound nodes.
+
+### The shape of the bound tree should correspond to the shape of the program's static semantics
+
+Generally speaking, that means that there is an isomorphism between the syntax and bound nodes, except when there is a mismatch between the shape of the two.  When possible, we prefer the correspondence be as direct as possible.  Here are two examples that illustrate this:
```
Comment (jcouv) [inline docs/compilers/Design/Bound Node Design.md]: nit: "except when there is a mismatch between the shape of the two" seems redundant

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 4 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: Fixing it and adding that to the bound node design document in the next iteration.

---
In reply to: [359482457](https://github.com/dotnet/roslyn/pull/40457#discussion_r359482457) [](ancestors = 359482457)

File: docs/compilers/Design/Bound Node Design.md (hunk 5 above)
Comment (gafter) [reply docs/compilers/Design/Bound Node Design.md]: Clarifying.  I meant when there is a mismatch between the syntax and semantics.

---
In reply to: [359572054](https://github.com/dotnet/roslyn/pull/40457#discussion_r359572054) [](ancestors = 359572054)

**Comment by gafter** [comment]
@AlekseyTs @jcouv I've addressed your comments.  Even though you've already signed off, you may be interested to see how.

**Comment by jcouv** [comment]
@gafter Ready to merge :-) Thanks

**Pull request merged**